### PR TITLE
Remove most version detection and conditionals for older versions of Rust

### PR DESCRIPTION
**Pull request by joshtriplett**
Per library team decision, drop most of the conditionals, simplifying the code
substantially. Update the README accordingly.

I left the conditional for 1.62, since that *just* shipped, and the conditional for 1.51 pending ratification of a libc MSRV policy.

- Require rust >= 1.15 and drop libc_priv_mod_use conditional
- Require rust >= 1.19 and drop libc_union conditional
- Require rust >= 1.24 and drop libc_const_size_of conditional
- Require rust >= 1.25 and drop libc_align conditional
- Require rust >= 1.26 and drop libc_int128 conditional
- Require rust >= 1.30 and drop libc_core_cvoid conditional
- Require rust >= 1.33 and drop libc_packedN conditional
- Require rust >= 1.33 and drop libc_cfg_target_vendor conditional
- Require rust >= 1.37 and drop libc_underscore_const_names conditional
- Require rust >= 1.40 and drop libc_non_exhaustive conditional
- Remove array size hacks for Rust < 1.47
- README.md: Update version-dependent features

Deferred while discussing libc MSRV policy:

- Require rust >= 1.51 and drop libc_ptr_addr_of conditional
- Require rust >= 1.62 and drop libc_const_extern_fn conditional

### CODE REVIEW

**Comment by Amanieu** [comment]
cc @JohnTitor

File: src/sgx.rs [hunk 1]
```diff
@@ -1,5 +1,7 @@
 //! SGX C types definition
 
+use c_void;
```
Comment (Amanieu) [inline src/sgx.rs:3]: Should this be `pub use` instead?

Comment (Amanieu) [reply src/sgx.rs:3]: Ah nevermind, I see there's one in the crate root.

File: src/unix/bsd/netbsdlike/openbsd/mod.rs [hunk 2]
```diff
@@ -815,135 +815,131 @@ cfg_if! {
...
+                    .field("f_favail", &self.f_favail)
+                    .field("f_syncwrites", &self.f_syncwrites)
+                    .field("f_syncreads", &self.f_syncreads)
+                    .field("f_asyncwrites", &self.f_asyncwrites)
+                    .field("f_asyncreads", &self.f_asyncreads)
+                    .field("f_fsid", &self.f_fsid)
+                    .field("f_namemax", &self.f_namemax)
+                    .field("f_owner", &self.f_owner)
+                    .field("f_ctime", &self.f_ctime)
+                // FIXME: .field("f_fstypename", &self.f_fstypename)
```
Comment (Amanieu) [inline src/unix/bsd/netbsdlike/openbsd/mod.rs:910]: Why is this left as a FIXME? Can we just use `#[derive(Debug)]` here?

**Comment by Kixunil** [comment]
I'm very disappointed hearing this. This is a significant jump in MSRV which not all parties can deploy quickly. As an example, Debian stable is now at 1.48 which is below 1.56. `libc` is one of the most important crates out there, pretty much required for FFI. Doing this pretty much kills Rust FFI for anyone who's not in the "go fast and break things" camp - pretty much all security-critical software.

**Comment by ibraheemdev** [comment]
It looks like 1.40 could be supported while still getting most of the benefit from this PR.

**Comment by Kixunil** [comment]
@Amanieu nothing wrong with bumping super-ancient 1.13. 1.40 would be perfect for me, but 1.48 is fine too.

**Comment by Noah-Kennedy** [comment]
@Amanieu I think it's probably time for libc to establish an MSRV policy from what I'm seeing here.

**Comment by Kixunil** [comment]
@Amanieu my preference would be "what popular Linux distros (insert list) support in repositories" I specifically care about Debian stable but there are other popular distros too.

**Comment by pfmooney** [comment]
> I think it's probably time for libc to establish an MSRV policy from what I'm seeing here.

Stepping back to consider other flag day conditions that `libc` might encounter could be worthwhile

**Comment by Noah-Kennedy** [comment]
> @Amanieu my preference would be "what popular Linux distros (insert list) support in repositories" I specifically care about Debian stable but there are other popular distros too.

So, I'm actually not so sure on this, as relying on debian stable will cause a huge maintenance burden over time as debian lags further and further behind due to the infrequent release cycle. I think some sort of time-based policy (e.g, last year's worth of releases) may be the best approach here as it keeps some stability but also doesn't lock in MSRV behind what debian is doing.

**Comment by Kixunil** [comment]
@Noah-Kennedy Debian tries to target 2 year release cycle. They would rather put off the release if there are unresolved problems but I don't think that's too bad. A few months shouldn't be a huge problem, I guess. If it gets unreasonably prolonged I wouldn't mind MSRV bump. So perhaps "whatever popular distros support unless they unreasonably prolong releases; minimum 2 years" :)

**Comment by Noah-Kennedy** [comment]
Hmm, two years sounds like a good timeline actually for MSRV breaks in libc.

**Comment by Noah-Kennedy** [comment]
Honestly, libc should be treated as something that doesn't do anything even mildly breaking outside of fixed release dates every few years. It is one of the core foundational elements of the rust ecosystem, so a change in MSRV here requires everyone to bump MSRV, which is a pretty big deal.

This needs to be done with caution, not everyone can update to the newest compiler for entirely valid reasons, and those who have this issue shouldn't be forced to rely on an old version of the rust ecosystem built on an older version of libc.

File: src/unix/bsd/netbsdlike/openbsd/mod.rs (hunk 2 above)
Comment (joshtriplett) [reply src/unix/bsd/netbsdlike/openbsd/mod.rs:910]: @Amanieu This is just moving code we already have from elsewhere. It should absolutely get fixed, but not as part of this change.

**Comment by joshtriplett** [comment]
> As an example, Debian stable is now at 1.48 which is below 1.56.

And Debian stable's packaged version of the libc crate is 0.2.80, which works on Debian stable's packaged version of rustc. Debian's rustc is for use by Debian packages. Compatibility with the rustc shipped by Debian *stable* is not a goal crates on crates.io should be striving to meet.

> I don't think supporting 1.48 costs us too much

Not at the moment, but the implied MSRV policy of "support whatever's in Debian stable" absolutely would long-term.

> As is, this will force a lot of ecosystem crates that make MSRV guarantees to bump their minimum versions.

I sincerely hope so. MSRV is a prisoner's dilemma: crates keep MSRV low because some other crate asks them to keep MSRV low because some other crate asks them to keep MSRV low, *most* of which can't actually point to a project that actually *needs* that MSRV for some reason. As a result, there are far *more* crates affected than there are users who actually have any desire to *use* such a Rust version.

We discussed possible MSRV policies in the library team meeting. The primary debate was between "support stable and stable-minus-one to allow time for upgrades" and "support a few extra versions back because we're libc", but the debate was between N-1 and (say) N-3 or N-4, not "1-2 years".

(To be clear, nobody proposed that we upgrade gratuitously, only that we upgrade when there's a rustc feature we need.)

We ended that meeting without setting a concrete MSRV policy, largely because we figured we'd start by making this change and go from there. But it looks like we may need to set a concrete MSRV policy after all.

> This needs to be done with caution, not everyone can update to the newest compiler for entirely valid reasons, and those who have this issue shouldn't be forced to rely on an old version of the rust ecosystem built on an older version of libc.

Yes, they absolutely should. If you run old rustc, run a matching set of old crates, such as those packaged alongside your rustc. Whoever packaged your rustc and your crates will then have validated that your rustc builds your crates.

**Comment by joshtriplett** [comment]
The concrete MSRV policy I would propose:

- Support the last three stable versions of rustc (stable and the prior two versions).
- Only upgrade if there's a feature we actually need.

**Comment by rtzoeller** [comment]
@joshtriplett I think the "Only upgrade if there's a feature we actually need" statement still leaves some ambiguity around what "need" really means, but perhaps that's intentional. I.e. if the lack of a language feature can be worked around, how expensive can that workaround be before the feature is "needed"?

Orthogonally, I'd request that a release be cut immediately prior to this PR being merged, and the same for any future MSRV bumps. IMHO a notice period before the MSRV is bumped would also be helpful - it'd give anyone who wants changes made to version of libc supporting a soon-to-be-dropped version of Rust a window to get them submitted.

**Comment by udoprog** [comment]
The way @joshtriplett describes the downstream use of MSRV largely echoes with me. Anyone has the ability to vendor their dependencies.

What MSRV policies across the ecosystem gives is a sliding window of released crates which makes vendoring a specific rustc release possible (in say the debian repository). But the minimal requirement is that the ecosystem *can* at some point be frozen on a reasonably stable set of package versions which allows everything being vendored to build. The wider the MSRV of all packages the better since it increases the probability that this global set of versions can be found (else package maintainers have to go through the hassle of figuring out how to have more than one version of the same package in parallel).

Another potentially large concern in my mind is having a bug patching policy which is friendly to downstreams, such as the one [outlined in Tokio](https://github.com/tokio-rs/tokio#bug-patching-policy). It serves to guide maintainers which minor version of a package to prefer to pick in the hopes that critical issues in the future will be patched and tested against a compatible version of the vendored package. A similar policy for foundational crates such as libc might be worth considering.

So far this is all a bit speculative and I'd love to hear more concrete user stories around this. Especially ones where people have been hurt by the lack of a stricter MSRV for some package with no reasonable remediation at hand.

**Comment by Darksonn** [comment]
The proposal of last three stable versions would force Tokio to break its MSRV policy, which is that the MSRV must always be at least six months old. (i.e. our policy is N-4)

The Tokio project controls most of the dependencies of Tokio, so until now, this is a policy that we've been able to uphold without issues. I see now that it was a mistake not to consult the libc maintainers before picking this policy. The reason this didn't happen is that libc is such a fundamental crate that I never imagined that libc would pick something weaker.

I would really appreciate it if the policy was N-4 rather than N-2.

---

As for debian stable, Tokio doesn't track debian stable, but there are every so often people who complain that we don't. People definitely try to use the cargo from the debian stable repos to compile their projects.

> I'd love to hear more concrete user stories around this. Especially ones where people have been hurt by the lack of a stricter MSRV for some package with no reasonable remediation at hand.

I do have a user story about this, though it was a case where the MSRV being too new was merely annoying and wasted my time, rather than a situation with no reasonable remediation.

The story is that I have a custom program I run in a cronscript on a VPS to manage my RSS feeds. Initially I wanted to compile it on my local machine and scp the binary to the server, but the binary didn't work on the server due to a version mismatch in the dynamic libraries (I think it was glibc in this case, but I've run into the same issue with an sql library in the past). I then tried to compile it on the server, but it didn't compile because Tokio's MSRV was too new. Thus, I was forced to install rustup to compile the application. This is annoying because I wanted Rust to be available to all users on the VPS. Looking at the VPS, it appears that I've somehow managed to install it globally, although it appears that I did it wrong, because upgrading rustc failed when I just tried it.

**Comment by Kixunil** [comment]
@joshtriplett perhaps it'll help to describe my use case. I write security-critical software that has to run on a well-secured system which should also be reliable. As such I chose Debian stable as the base system. In principle I have only a single reason against picking whichever version of Rust: it's hard to ensure the installed version is not backdoored unless it came from distribution repository.

The alternative for me would be to make effort to make Rustup verify Rust signatures when installing (last time I checked it didn't do that) and then get legit Rustup. Or write my own alternative installer.

Packaged crates don't fit my use case well mainly because there's no way to tell `cargo` to pull crates from system repository unless they don't exist there. Note that I'm not forgetting crate security. After reviewing a crate I can just rely on its hash being in `Cargo.lock`, yes, I always use `--locked`.

So far sticking with MSRV was the solution for me but since it's getting harder over time, I may attempt to do something different.

**Comment by tshepang** [comment]
One option is for this to be gradual, where, for example, each release bumps a requirement. For example,  libc 0.2.127 requires minimum Rust 1.15, then libc 0.2.128 requires minimum Rust 1.19, and so on. I feels it will be less disruptive this way, and less controversial.

File: build.rs [hunk 3]
```diff
@@ -18,6 +17,9 @@ fn main() {
              please consider using the `std` cargo feature instead\""
         );
     }
+    if env::var("CARGO_FEATURE_ALIGN").is_ok() {
+        println!("cargo:warning=\"libc's align cargo feature is deprecated (and always available)");
```
Comment (CAD97) [inline build.rs:21]: (Just an aside: `cargo:warning` messages are [only emitted for local/path dependencies](https://doc.rust-lang.org/cargo/reference/build-scripts.html#cargo-warning) if `-vv` is not used, so these warnings are unlikely to ever be seen.)

**Comment by mati865** [comment]
@Darksonn you don't have to require latest `libc` version in Tokio.
If somebody uses Rust vendored by distro they should also use dependencies vendored by distro or manually pin crates to older versions with `Cargo.lock`.

**Comment by cuviper** [comment]
> * Require rust >= 1.56 so we can use `rust-version`

You don't *need* 1.56 for `rust-version` -- it's ignored on earlier versions, but still serves as documentation.

```
warning: `rust-version` is not supported on this version of Cargo and will be ignored
```

File: build.rs (hunk 3 above)
Comment (joshtriplett) [reply build.rs:21]: That's a good point, thank you.

**Comment by joshtriplett** [comment]
@cuviper Ah, fair enough. I had assumed that including `rust-version` would cause old versions to fail to build anyway due to the unknown key, and thus it was a choice between omitting it entirely or setting it to at least the version that supported `rust-version`. But you're right, it's a warning, not an error.

**Comment by joshtriplett** [comment]
With the most recent change, this PR now leaves libc still (currently) working with a 15-month-old version of Rust (N-11). That's *not* a commitment to *continue* supporting versions that old, and we still absolutely need an official policy; I'd love to not reopen this discussion *every time* we use a new feature. But for the purposes of *this* PR, this is now exceeding the support window of any and all of the plausible policies that seem even remotely likely to be considered in https://github.com/rust-lang/libs-team/issues/72 .

Thus, I'm hoping we can go ahead and merge this, before it starts getting merge conflicts, while continuing to discuss policy for future changes in the other issue without losing momentum on that.

**Comment by rtzoeller** [comment]
@joshtriplett would it be possible to get a release cut immediately before this is merged? It'd be nice to have code that's already been merged make it into a release with the previous MSRV.

**Comment by carllerche** [comment]
Tokio's current LTS release which [currently supports Rust 1.49](https://github.com/tokio-rs/tokio/blob/tokio-1.18.x/.github/workflows/ci.yml#L16) up until at a minimum January 2023. I understand that, in theory, we *=could ask our users on the LTS release to pin their dependencies, but I don't think that meets the spirit of the LTS version.

**Comment by taiki-e** [comment]
Removing the commit to increase MSRV to 1.51 (https://github.com/rust-lang/libc/pull/2845/commits/8dd8ce5cc530a600d77ef6bcd3fc16148f15f884) seems to meet most of the requests. (Removal of annoying hacks like the array hack from libc, support debian stable, support tokio MSRV, etc.)

**Comment by joshtriplett** [comment]
@carllerche We have to draw a line somewhere, and I was quite serious when I said "exceeding the support window of any and all of the plausible policies that seem even remotely likely to be considered". The end state of "but if you extended it a bit *more* that would make it even less work for *our* support commitment to *our* customers" is supporting Rust 1.0 (and yes, I've heard from people who think we should do that, as well as from people who think "C code shouldn't require C11, that's too new").

@taiki-e The only reason I dropped the requirement for 1.56 is that it turned out to not provide any value at all. This issue is going to arise *again* for the next feature, and the next, and the next. Continuing to support 1.47 is just kicking the can down the road. (And no, the difference between 1.47 and 1.51 does not meet *all* of the requests, any more than 1.51 does.) There is *no* solution that will meet all desires here, just tradeoffs to be made.

**Comment by cuviper** [comment]
> The end state of "but if you extended it a bit _more_ that would make it even less work for _our_ support commitment to _our_ customers" is supporting Rust 1.0

I don't think it's necessary to invoke a slippery slope here. This PR is already taking a **massive** step forward from 1.13, and it seems perfectly reasonable to consider a slightly smaller step, especially when you've emphasized that this is not setting future policy.

IOW, the ask is not really about *extending*, rather to not *shrink* the status quo quite so much.

**Comment by seanmonstar** [comment]
> I'm hoping we can go ahead and merge this, before it starts getting merge conflicts, while continuing to discuss policy for future changes in the other issue without losing momentum on that.

As a point from an outsider, since it seems that the changes in this PR change an unwritten policy, and the policy hasn't been settled in the related issue yet, merging this first seems premature. I wouldn't think that the risk of merge conflicts would be reason to skip settling the policy.

**Comment by carllerche** [comment]
It is also worth noting that libc's MSRV policy is [stated in the README](https://github.com/rust-lang/libc/#rust-version-support) and users *are* counting on it.

**changing** the policy does seem to merit a breaking change release.

**Comment by cuviper** [comment]
I don't agree that changing MSRV should be a semver break, but FWIW, cargo docs call this "possibly-breaking".
https://doc.rust-lang.org/cargo/reference/semver.html#env-new-rust

**Comment by carllerche** [comment]
@cuviper

> I don't agree that changing MSRV should be a semver break, but FWIW, cargo docs call this "possibly-breaking". https://doc.rust-lang.org/cargo/reference/semver.html#env-new-rust

I'm not saying all MSRVs should issue a breaking semver when dropping older MSRVs, what I am saying is `libc` has **explicitly** stated that it supports Rust 1.13.0. Consumers of libc (e.g. Tokio). have made decisions based on this stated policy. Changing that policy would break consumers that have made assumptions using that policy. It seems that, if libc wants to **change** its stated policy to N-?, then *that* should be w/ a breaking semver release and then support for older rustcs can be dropped **without** a breaking change following the new policy.

**Comment by joshtriplett** [comment]
@cuviper wrote:
> it seems perfectly reasonable to consider a slightly smaller step

There is a meaningful delta here between "stepping forward to a version that represents actual changes" and "not setting an MSRV newer than any actual changes we need". I was happy to drop the 1.56 requirement because it served absolutely no effect. If we start counting "and how many lines of code would *this* version support require", that's a line I'd much rather *not* cross.

**Comment by CAD97** [comment]
On the other hand, libc documented its supported toolchain version before it was common practice to document "msrv policy" and "msrv actual" separately. Having not changed this does defacto mean that its documented as both, but it can be argued that the policy for upgrading the supported msrv was undocumented.

But using that argument, I would **heavily** suggest not changing the supported msrv without documenting the policy separately from the actual msrv. The argument that the policy was not documented only works once.

**Comment by taiki-e** [comment]
@joshtriplett

> This issue is going to arise _again_ for the next feature, and the next, and the next. Continuing to support 1.47 is just kicking the can down the road. (And no, the difference between 1.47 and 1.51 does not meet all of the requests, any more than 1.51 does.)

I did not mean to suggest that MSRV be pinned forever.
My point is that 1.47, 1.48, and 1.49 are version requirements that have been specifically mentioned several times in this thread or the libs-team thread, and it seems easy to retain support for them for now.
I felt that retaining those supports for now was a reasonable compromise to continue to discuss the MSRV policy while reducing maintenance costs and conflict in communicating with the community.

**Comment by BurntSushi** [comment]
@carllerche I think the implication of your interpretation here is that the `libc` README is specifically providing a semver guarantee that `libc 0.2.x` will compile on Rust 1.13 for all time, right? That wording got merged in #1254, and I'm not seeing anything that would lead me to believe that was what was intended. The wording also doesn't seem to explicitly say that either... @CAD97 is right. The wording doesn't tell readers whether it's descriptive or prescriptive.

@joshtriplett I can appreciate wanting to get some version of this PR merged. IMO, if there's push back from stakeholders, then we should block said change on answering the policy question first. But if folks are happy in practice with, say, bumping the MSRV to Rust 1.49, then that seems like a fine thing to do while we discuss the broader policy. I would imagine it's still a nice win that will let us delete a lot of conditional compilation.

**Comment by joshtriplett** [comment]
@BurntSushi If we have a consensus from the team that we're willing to merge this if it rolls back the 1.51 requirement, and not willing to merge this if it doesn't, then I can roll that back and open another PR for the remainder of the change. Do we have that consensus?

**Comment by BurntSushi** [comment]
@joshtriplett 

> If we have a consensus from the team that we're willing to merge this if it rolls back the 1.51 requirement, and not willing to merge this if it doesn't

I think that's my position, yes. At least until we settle the policy question.

Whether this reflects a consensus, I'm not sure. :-) I think we'd need to hear from others. cc @rust-lang/libs-api 

Let me try to make a succinct statement of what we'd like to decide here:

* https://github.com/rust-lang/libs-team/issues/72 is where we are currently trying to decide the MSRV policy of libc. It seems we are at a bit of an impasse there.
* While we work to resolve the impasse, it's still helpful to bump libc's MSRV. Its current MSRV, 1.13, is very very old. The question is, what do we bump it to?
* As someone following https://github.com/rust-lang/libs-team/issues/72, my read of the situation is that there is nobody advocating in favor of maintaining a 1.13 MSRV. So it should be greater than 1.13.
* There are multiple folks from distinct projects who have been advocating in favor of anywhere from 1 to 2 (I think) years of MSRV support. So that's somewhere around N-9 to N-18, which gives us a range of Rust 1.44 on the most conservative end to Rust 1.53 on the "still conservative, but not universally appealing" end. (Others have of course argued for an MSRV <1 year, but I'm limiting my scope to the conservative viewpoint given that the policy question is still being decided.)
* In this thread, folks have mentioned that Rust 1.53 would be a burden for them.
* I've suggested that we take the win that we can get. If there aren't any objections, I'd propose whatever is in current Debian stable as a benchmark. I think that's **Rust 1.48**. I don't think I've seen anyone earnestly argue for support of an older version of that. (But I could be wrong, there has been a ton of discussion.)
* To be explicit: the action we take here shouldn't serve as a precedent for the general MSRV policy question. The idea here is to try to improve the interim situation in a way that pretty much everyone is happy-ish with.

**Comment by CAD97** [comment]
(I'm not T-libs, but I would request still blocking such a PR on explicitly noting in the readme that the MSRV note is descriptive only, and the policy is being discussed, with a link. And being ready to receive complaints about the MSRV change to that link, because any change *will* break someone. But 1.47 (21 months) does seem a safe baseline per available discussion here in a way 1.51 (16 months) doesn't quite.)

**Comment by joshtriplett** [comment]
@CAD97 That's a really good idea. (I don't think it's appropriate to link to a GitHub issue from the README, but I've linked it from the PR.)

https://github.com/rust-lang/libc/pull/2852

**Comment by Kixunil** [comment]
> I'd propose whatever is in current Debian stable as a benchmark. I think that's **Rust 1.48.**

Correct, it is.

**Comment by workingjubilee** [comment]
I feel like taking Debian Stable as a benchmark even this one time will still be precedent-suggestive. As epage, the author of Clap, noted, in practice users can be keenly attentive to de facto policies and may plan around them, not de jure.

It also has the issue that Debian does not follow a steady release cadence for Rust, and what cadence they do have **falls behind**. Their previous releases seemingly followed a 7 rustc version cadence (7... for a ~2 year cycle, which would be about 17 normally! so each Debian has been slipping roughly 10 versions behind...). Currently they ship rustc 1.59 for their will-be-stable-next-year release, which breaks that cadence by jumping from rustc 1.48 to there, or 11 versions. Debian says they'll update it before the freeze, presumably to 1.62.1 at minimum, which would make the gap even wider, at least 14 versions! But then what about their next release after that? The ball is entirely in their court. They might ship rustc 1.63 instead of jumping more versions again, maybe, in 2025.

Meanwhile, Ubuntu just ships rustc 1.59 to active Ubuntu installations, from Debian's testing packages! I guess we're that reliable? And as I noted in the IRLO thread, Debian Stable will release a "rustc-mozilla", with a different version, to use for building certain packages instead if they are genuinely pressed to ship a patch for Mozilla software, like Firefox.

And several people in the MSRV policy thread have noted that we can only really see if the version change in bumping libc causes problems for people by, well, doing it and then listening for the screams. We expect to hear some specific negative feedback somewhere between an MSRV bump of 1.34~1.62 (indeed several people have mentioned they already anticipate it, obviously), but it will be hard to isolate that signal if we do it all in one single step.  It would also mean that the obvious revert of this PR, if we decided we must revert, would push us all the way back to the start. That would seem disastrous, as would debating which parts to keep all over again, essentially un-resolving whatever resolution, even partial, we have arrived at.

I propose instead that we attempt to ship this entire changeset... in phases.
This PR included, originally, ~13 edition-based changes here, moving up ~43 editions.
To take the midpoint of the changes would advance us to 1.33.
To take the midpoint of the editions would advance us to 1.33~1.37.

From there we can ratchet the bisection, with a libc release for each one:
- 1.33 (the all important `repr(packed)`, `cfg(target_vendor = "debian")`, and `ffi::c_void`)
- 1.47 (`non_exhaustive` and "real" arrays)
- 1.51 (`ptr::addr_of!`)
- 1.56 (rustc-version)

We can give each the space of about a rustc release cycle (so the changes to libc's rustc minimum have time to propagate), and then calibrate our next steps after each release based on the feedback. I expect any of the last three might have people pushing for a rollback. I am... expecting the first one is virtually completely safe. So far, virtually no one has seriously requested an MSRV policy that would block it, even on the ultra-conservative end, so I think we can make that change. But I have seen a few intimations that suggest jumping to 1.47 right away might provoke controversy.

If the goal is "let's take the win we can get", I think we should find *exactly* how much "free space" we can claim. Collapsing on Debian Stable's current release as the Schelling point seems like just setting a policy without actually writing it down. It may seem like this essentially means that we are handing out invitations to people who have already objected to then show up and formally object again in about 12~18 weeks, but that's what kicking this can down the road actually would look like. We can save a few extra joysticks in advance for our Street Opiner: World Posting Championship tournament.

**Comment by joshtriplett** [comment]
In the interests of getting most of this merged, I've updated this PR to only require 1.47, which is N-15 (nearly N-16) at this point, and older than the version currently in Debian stable. This is *not* a proposal or precedent for an MSRV policy to continue supporting versions that old indefinitely; rather, this is an attempt to incrementally improve the situation, on the basis of what the vast majority of people don't seem inclined to argue against.

@workingjubilee I think we can successfully move forward incrementally without treating the first point we move to as precedent. (Also, 1.56 turns out to not have anything we need; 1.51 and 1.62 are the next points with things we need.)

cc @Amanieu

**Comment by joshtriplett** [comment]
I've also filed https://github.com/rust-lang/libc/pull/2859 for a 0.2.127 release. That'll make 0.2.128 the release with an updated MSRV; nice round number, easy to remember. :smile:

**Comment by Amanieu** [comment]
You need to remove the CI jobs that check older Rust versions.

**Comment by joshtriplett** [comment]
@Amanieu Ah, oops. I thought I had, but apparently not.

**Comment by joshtriplett** [comment]
@Amanieu Fixed, hopefully.

**Comment by joshtriplett** [comment]
(Not sure why the aarch64 build for Windows is failing. Will investigate.)

**Comment by Amanieu** [comment]
```
   --- stderr
  error: expected identifier, found `_`
     --> ../src/fixed_width_ints.rs:64:23
      |
  64  |                   const _: [(); $a] = [(); $b];
      |                         ^
      | 
     ::: ../src/lib.rs
      |
  70  | / cfg_if! {
  71  | |     if #[cfg(windows)] {
  72  | |         mod fixed_width_ints;
  73  | |         pub use fixed_width_ints::*;
  ...   |
  133 | |     }
  134 | | }
      | |_- in this macro invocation
      |
      = note: `_` is a wildcard pattern, not an identifier
```

File: .github/workflows/bors.yml [hunk 4]
```diff
@@ -204,11 +204,6 @@ jobs:
           stable,
           beta,
           nightly,
-          1.13.0,
-          1.19.0,
-          1.24.0,
-          1.25.0,
-          1.30.0,
```
Comment (cuviper) [inline .github/workflows/bors.yml:211]: Shouldn't there still be CI for the remaining MSRV? And maybe also the further version steps that are still detected in the build script -- just 1.51?

**Comment by joshtriplett** [comment]
> ```
>    --- stderr
>   error: expected identifier, found `_`
>      --> ../src/fixed_width_ints.rs:64:23
>       |
>   64  |                   const _: [(); $a] = [(); $b];

This should work on Rust 1.37 and newer, and that log *seems* to be saying it has a newer version of Rust than that (which it should). Trying to find why it has a version that doesn't support that.

File: .github/workflows/bors.yml (hunk 4 above)
Comment (joshtriplett) [reply .github/workflows/bors.yml:211]: @cuviper I don't think we need the latter, but yes, I'll add 1.47 there.

**Comment by joshtriplett** [comment]
Ah, I think I figured it out. Despite looking like a Rust error, this is being parsed by `garando_syntax`, which doesn't understand this syntax.

File: .github/workflows/bors.yml (hunk 4 above)
Comment (cuviper) [reply .github/workflows/bors.yml:211]: Pardon me for unresolving, but AFAICT you haven't added 1.47 to CI yet.

**Comment by joshtriplett** [comment]
@Thomasdezeeuw This particular change is allowing for Rust versions more than a year old. What is socket2's compatibility story, and what do its users actually need? (Ideally, responses should probably go in the other discussion issue.)

**Comment by Thomasdezeeuw** [comment]
> @Thomasdezeeuw This particular change is allowing for Rust versions more than a year old. What is socket2's compatibility story, and what do its users actually need? (Ideally, responses should probably go in the other discussion issue.)

The MSRV for Socket2 is 1.46 (https://github.com/rust-lang/socket2#minimum-supported-rust-version-msrv) this pr increases it to 1.47 (according to the change in the README). We are in the process of increasing it to 1.63, but that will be done in v0.5 and might take a while.

Also I see a lot of Rust versions were removed from the CI, but the MSRV wasn't added.

**Comment by joshtriplett** [comment]
On Wed, Aug 10, 2022 at 05:07:23AM -0700, Thomas de Zeeuw wrote:
> Also I see a lot of Rust versions were removed from the CI, but the MSRV wasn't added.

Done locally, just not pushed yet. Will still fail CI for the moment due
to https://github.com/JohnTitor/ctest2/issues/38

**Comment by asomers** [comment]
Is there any way to help push this through?  I'd really like to finish it, because it's blocking #2813 .

**Comment by ChrisDenton** [comment]
Look at the list, I think the MSRV could be bumped to 1.33 without patching ctest2?

**Comment by tgross35** [comment]
https://github.com/rust-lang/libc/pull/4040 raised the MSRV to 1.63. I was able to cherry pick most of these commits to remove the config, with the exception of those that we need to somehow patch `ctest` for (thanks Josh for doing the work :) ).

**Pull request closed without merging**